gocourse18/dbwriter.go

```diff
@@ -0,0 +1,119 @@
+package main
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/jmoiron/sqlx"
+	_ "github.com/lib/pq"
+)
+
+type DBWriter struct {
+	db *sqlx.DB
+}
+
+func initDBWriter(ctx context.Context, dbname string) (*DBWriter, error) {
+	db, err := sqlx.ConnectContext(ctx, "postgres", "host=localhost port=5432 user=user password=password dbname=postgres sslmode=disable")
+	if err != nil {
+		return nil, fmt.Errorf("cannot connect to db: %v", err)
+	}
+
+	dbw := &DBWriter{
+		db: db,
+	}
+
+	if dbname != "postgres" {
+		err = dbw.createDatabaseIfNotExists(ctx, dbname)
+		dbw.db.Close()
+		if err != nil {
+			return nil, err
+		}
+
+		dbw.db, err = sqlx.ConnectContext(ctx, "postgres", fmt.Sprintf("host=localhost port=5432 user=user password=password dbname=%s sslmode=disable", dbname))
+		if err != nil {
+			return nil, fmt.Errorf("cannot connect to db: %v", err)
+		}
+	}
+
+	return dbw, nil
+}
+
+func (w *DBWriter) createDatabaseIfNotExists(ctx context.Context, dbName string) error {
+	var existingDBName string
+	err := w.db.QueryRowContext(ctx, "SELECT datname FROM pg_database WHERE datname = $1", dbName).Scan(&existingDBName)
```

Comment:
Здається, можна спростити трішки:
```suggestion
	err := w.db.QueryRowContext(ctx, "SELECT datname FROM pg_database WHERE datname = $1", dbName).Err()
```